src/main/kotlin/com/exactpro/th2/rptdataprovider/services/grpc/GrpcService.kt

```diff
@@ -0,0 +1,159 @@
+                    it.completableDeferred.complete(response)
+                }
+                    ?: LOGGER.trace {
+                        val firstSequence =
+                            decodedBatch.groupsList.firstOrNull()?.messagesList?.firstOrNull()?.sequence
+                        val lastSequence =
+                            decodedBatch.groupsList?.lastOrNull()?.messagesList?.lastOrNull()?.sequence
+                        val stream =
+                            "${decodedBatch.groupsList.firstOrNull()?.messagesList?.firstOrNull()?.message?.sessionAlias}:${decodedBatch.groupsList.firstOrNull()?.messagesList?.firstOrNull()?.message?.direction.toString()}"
+                        "codec response with hash ${response.responseHash} has no matching requests (stream=${stream} firstId=${firstSequence} lastId=${lastSequence} requestId=${response.requestId})"
+                    }
+            }
+        }
+
+        override fun onError(t: Throwable?) = LOGGER.error(t) { "gRPC request failed." }
+        override fun onCompleted() {}
+    }
+
+    override suspend fun sendToCodec(request: CodecBatchRequest): CodecBatchResponse {
+
+        return withContext(requestSenderScope.coroutineContext) {
+            while (pendingRequests.keys.size > maximumPendingRequests) {
+                delay(100)
+            }
+
+            pendingRequests.computeIfAbsent(request.requestId) {
+                val pendingRequest = request.toPending()
+
+                callbackScope.launch {
+                    delay(responseTimeout)
+
+                    pendingRequest.completableDeferred.let {
+                        if (it.isActive &&
+                            pendingRequests[request.requestId]?.completableDeferred == pendingRequest.completableDeferred
+                        ) {
+
+                            pendingRequests.remove(request.requestId)
+                            it.complete(null)
+
+                            codecLatency.gaugeDec(listOf(request.streamName))
+                            codecLatency.setDuration(
+                                pendingRequest.startTimestamp.toDouble(),
+                                listOf(request.streamName)
+                            )
+
+                            LOGGER.warn {
+                                val firstSequence =
+                                    request.protobufRawMessageBatch.groupsList.first()?.messagesList?.first()?.rawMessage?.sequence
+                                val lastSequence =
+                                    request.protobufRawMessageBatch.groupsList.last()?.messagesList?.last()?.rawMessage?.sequence
+                                val stream =
+                                    "${request.protobufRawMessageBatch.groupsList.first()?.messagesList?.first()?.rawMessage?.sessionAlias}:${request.protobufRawMessageBatch.groupsList.first()?.messagesList?.first()?.rawMessage?.direction.toString()}"
```

Comment:
Same as above. Also why do we use `first`/`last` here, but `firstOrNull`/`lastOrNull` above?